ghcide/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -86,7 +125,43 @@ instance NFData ShowDiagnostic where
 --   along with the related source location so that we can display the error
 --   on either the console or in the IDE at the right source location.
 --
-type FileDiagnostic = (NormalizedFilePath, ShowDiagnostic, Diagnostic)
+data FileDiagnostic = FileDiagnostic
+  { fdFilePath :: NormalizedFilePath
+  , fdShouldShowDiagnostic :: ShowDiagnostic
+  , fdLspDiagnostic :: Diagnostic
+  , fdStructuredMessage :: StructuredMessage
+  }
+  deriving (Eq, Ord, Show, Generic)
+
+instance NFData FileDiagnostic
+
+data StructuredMessage
```

Comment:
This is really only a GHC-specific message. Is the idea that we might expand it to other types with some `Typeable` wizardry or something? If so, write it down!

Reply:
The idea behind `StructuredMessage` is that it's a thin wrapper around `GhcMessage` that pretends the type has Show, Eq, Ord, Generic instances. This allows us to derive those typeclasses. The fact that it isn't a law-abiding typeclass is not important since it's just metadata in `FileDiagnostic`.

Documented in 003c15c1

This is a bit cursed so I'm leaving this open - wdyt?

Reply:
I guess my question is: why do we need those instances? And if we need them for a good reason, area we sure the use is safe if the instances are busted? I think the answer is "yes it's okay, because the LSP diagnostic should always 'determine' the `StructuredMessage`, so we can safely ignore it", but we should write that reasoning down!